### PR TITLE
Add loader and storage support for struct

### PR DESCRIPTION
**Pull request by acquamarin**
This PR adds basic loader and storage support for struct dataType.
Unresolved struct dataType issues:
1. Null struct value and null struct field value.
2. Improves the loader for struct so that we can support more field types.
3. Solves `Match (m:movies) return m.description.rating` exception bug.
4. Query processing for unflat/flat struct field valuevectors.
5. Implements lookup/write API for struct column.

### CODE REVIEW

File: src/storage/in_mem_storage_structure/in_mem_column_chunk.cpp [hunk 1]
```diff
@@ -127,5 +145,139 @@ void InMemColumnChunk::setValueFromString<common::timestamp_t>(
...
+            reinterpret_cast<common::StructTypeInfo*>(dataType.getExtraTypeInfo());
+        auto structFieldTypes = structTypeInfo->getChildrenTypes();
+        auto structFieldNames = structTypeInfo->getChildrenNames();
+        std::regex whiteSpacePattern{"\\s"};
+        // Removes the leading and trailing '{', '}';
+        auto structString =
+            std::string(structView.data(), structView.length()).substr(1, structView.length() - 2);
+        std::istringstream structStream{structString};
+        std::string structField;
+        while (getline(structStream, structField, ',')) {
```
Comment (ray6080) [inline src/storage/in_mem_storage_structure/in_mem_column_chunk.cpp:179]: Let's see if you can avoid using `istringstream` and create new string here.

File: src/storage/in_mem_storage_structure/in_mem_column_chunk.cpp [hunk 2]
```diff
@@ -73,6 +75,22 @@ void InMemColumnChunk::templateCopyValuesToPage<std::string, InMemOverflowFile*,
...
+void InMemColumnChunk::templateCopyValuesToPage<std::string, common::CopyDescription&,
+    common::offset_t>(const PageElementCursor& pageCursor, arrow::Array& array, uint64_t posInArray,
+    uint64_t numValues, common::CopyDescription& copyDesc, common::offset_t nodeOffset) {
+    copyStructValueToFields(array, posInArray, copyDesc, nodeOffset, numValues);
+}
+
+template<>
+void InMemColumnChunk::setValueFromString<bool>(
+    const char* value, uint64_t length, common::page_idx_t pageIdx, uint64_t posInPage) {
+    std::istringstream boolStream{std::string(value)};
```
Comment (ray6080) [inline src/storage/in_mem_storage_structure/in_mem_column_chunk.cpp:88]: Ditto. Can we not use `istringstream` here?

**Review by ray6080 — APPROVED** [review]
Can you add more tests to cover more data types? See codecov report in `InMemStructColumnChunk::copyValueToStructColumnField`.

File: src/common/types/types.cpp [hunk 3]
```diff
@@ -51,15 +50,18 @@ uint64_t SerDeser::deserializeValue(StructTypeInfo& value, FileInfo* fileInfo, u
 }
 
 template<>
-uint64_t SerDeser::serializeValue(const StructField& value, FileInfo* fileInfo, uint64_t offset) {
-    offset = serializeValue(value.getName(), fileInfo, offset);
-    return serializeValue(*value.getType(), fileInfo, offset);
+uint64_t SerDeser::serializeValue(
+    const std::unique_ptr<StructField>& value, FileInfo* fileInfo, uint64_t offset) {
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/common/types/types.cpp:54]: StructFields are stored in a vector. The function to serialize/deserialize a vector passes the `unique_ptr<StructField>&` to the serilalizer/deserializer.

File: src/common/types/types.cpp [hunk 4]
```diff
@@ -170,6 +172,14 @@ std::vector<std::string> StructTypeInfo::getChildrenNames() const {
     return childrenNames;
 }
 
+std::vector<StructField*> StructTypeInfo::getStructFields() const {
+    std::vector<StructField*> structFields;
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/common/types/types.cpp:176]: DONE

File: src/common/types/types.cpp
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/common/types/types.cpp:1]: Removed typeID from the constructor

File: src/common/types/types.cpp [hunk 5]
```diff
@@ -149,7 +151,7 @@ std::unique_ptr<StructField> StructField::copy() const {
 std::unique_ptr<ExtraTypeInfo> StructTypeInfo::copy() const {
     std::vector<std::unique_ptr<StructField>> structFields;
     for (auto& field : fields) {
-        structFields.emplace_back(field->copy());
+        structFields.push_back(field->copy());
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/common/types/types.cpp:154]: done

File: src/common/types/types.cpp [hunk 6]
```diff
@@ -41,8 +41,7 @@ uint64_t SerDeser::deserializeValue(FixedListTypeInfo& value, FileInfo* fileInfo
 template<>
 uint64_t SerDeser::serializeValue(
     const StructTypeInfo& value, FileInfo* fileInfo, uint64_t offset) {
-    auto childrenTypesToReturn = value.getChildrenTypes();
-    return serializeVector(childrenTypesToReturn, fileInfo, offset);
+    return serializeVector(value.fields, fileInfo, offset);
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/common/types/types.cpp:44]: The templated serialize function signature requires that the data to serialize/deserailize must be a reference other than a pointer.

File: src/storage/in_mem_storage_structure/in_mem_column_chunk.cpp (hunk 1 above)
Comment (acquamarin) [reply src/storage/in_mem_storage_structure/in_mem_column_chunk.cpp:179]: Replaced it with `StringUtils::split`

**Pull request merged**